### DISCUSSION TITLE
Update Documentation / Playground for `draw` functions

### DISCUSSION THREAD

**Discussion by SHogenboom** [replit/kaboom#664] [Q&A]
All methods are intuitively named, and so when looking to draw a rectangle I figured the `drawRect` would do the job. However, I could not get the code to work. Not my own code, nor the examples in the documentation.

When browsing through the playground I saw a rectangle in the `gravity` demo. In this demo the `add([rect(x,y)]` method is used rather than the `drawRect`. From earliers Discord discussions I saw comments of people being unable to make the code work, with the “solution” being to use them in an `onDraw` method. Those suggestions were too cryptic for me to get everything in working order.

The same issue happened when comparing `drawText` with `add([text()]` - in this case the font did not exist in the `drawText` method, but worked perfectly fine in the `add([text()]` method.

I would suggest an update of the documentation to (1) show how the `drawRect` function should be used (in `onDraw`?), and (2) a suggestion to use the `add[(…)]` options instead.

**Comment by SHogenboom** [comment]
To elaborate on myself a minimal example comparing usage of `drawRect` and `add[(rect())]`.

**The Result**
![Screenshot 2022-12-31 at 16 09 31](https://user-images.githubusercontent.com/28568743/210141195-d3097dc6-5140-4b31-a5b9-424b50fb0126.png)

**The Code**
```
import kaboom from "kaboom";

// GLOBAL VARIABLES
const gameWidth = 500;
const gameHeight = 200;
const posX = 30; // each rect is 20 wide + some spacing
const posY = 50; // plot all on the same height

// CREATE GAME
kaboom({
  width: gameWidth,
  height: gameHeight,
});

// --------------------
// Example from Kaboom.js Documentation on `rect()`
// WORKS
// --------------------

add([pos(posX * 1, posY), rect(20, 40), outline(4), area()]);

// --------------------
// Example from Kaboom.js Documentation on `drawRect()`
// DOES NOT WORK
// No errors - Nothing Shown
// --------------------
drawRect({
  width: posX * 2,
  height: posY,
  pos: vec2(20, 20),
  color: YELLOW,
  outline: { color: BLACK, width: 4 },
});

// --------------------
// Adaptation with named parameters
// DOES NOT WORK
// No errors - Nothing Shown
// --------------------
add([pos({ x: posX * 3, y: posY }), rect(20, 40), outline(4), area()]);

// --------------------
// Adaptation with radius
// DOES NOT WORK
// No errors - Normal Rect
// --------------------
add([pos(posX * 4, posY), rect(20, 40, 10), outline(4), area()]);

// --------------------
// Adaptation with different color outline (alias)
// WORKS
// --------------------
add([pos(posX * 5, posY), rect(20, 40), outline(4, YELLOW), area()]);

// --------------------
// Adaptation with different color outline (rgb)
// DOES NOT WORK
// No errors - normal rect - not the orange border
// --------------------
add([pos(posX * 6, posY), rect(20, 40), outline(4, (255, 127, 0)), area()]);

// --------------------
// Adaptation with different color outline (rgb via named parameters)
// DOES NOT WORK
// No errors - rect without border
// --------------------
add([
  pos(posX * 7, posY),
  rect(20, 40),
  outline({ width: 4, color: (255, 127, 0) }),
  area(),
]);
```

Not in this minimal example - but the `layer()` method does not work for `add([rect()])`.

**Reply by hacknug** [reply]
Not sure how you're using the `layer()` component but it should work when used like this: https://kaboomjs.com/#layers

First you define the `layers()` for your game, and then you assign each object you create to one of those layers using `add([rect(), layer('game')])`.

Hope this helps.

**Comment by hacknug** [comment]
I think this is already mentioned on the docs: https://kaboomjs.com/#Draw

From that section:
> Also note that you have to put drawXXX() functions inside an onDraw() event or the draw() hook in component definitions which runs every frame (after the update events), or it'll be immediately cleared next frame and won't persist.

I also tried using `drawXXX()` methods once but later realised they we're not the right tools for the job I had in mind and haven't used them for anything yet.

**Comment by slmjkdbtl** [comment]
Yes since kaboom clears the canvas automatically every frame `drawXXX` need to be called every frame like in `onDraw()` to draw something persist.

You can think of `drawXXX()` as a lower level API than the components, internally all components uses `drawXXX()` function to render, like the [`rect()`](https://github.com/replit/kaboom/blob/dd6a97626ab3963341b32fc615486258a4226655/src/kaboom.ts#L4944-L4956) component just calls `drawRect()` every frame in its `draw` event which runs every frame.

`drawXXX()` functions are like the draw functions in p5 or canvas 2d, the only difference is kaboom internally clears the canvas every frame
